### PR TITLE
Improve code coverage

### PR DESCRIPTION
**Pull request by fabiojb**
Added some tests to improve code coverage. I intend to add more tests, so I want to know if it's better to add all tests in one PR or to open multiple minor PRs.

### CODE REVIEW

File: src/test/java/de/siegmar/logbackgelf/pool/SimpleObjectPoolTest.java [hunk 1]
```diff
@@ -93,6 +97,79 @@ public void maxLastBorrowed() throws InterruptedException {
...
+        final long before = System.currentTimeMillis();
+
+        new Thread(() -> {
+            try {
+                Thread.sleep(200);
+            } catch (InterruptedException e) {
+                throw new RuntimeException(e);
+            }
+            pool.returnObject(o1);
+        }).start();
```
Comment (osiegmar) [inline src/test/java/de/siegmar/logbackgelf/pool/SimpleObjectPoolTest.java:134]: This could be replaced by

```
Executors.newScheduledThreadPool(1)
    .schedule(() -> pool.returnObject(o1), 200, TimeUnit.MILLISECONDS);
```

**Review by osiegmar — CHANGES_REQUESTED** [review]

**Comment by osiegmar** [comment]
Thanks for your work on this!

> I intend to add more tests, so I want to know if it's better to add all tests in one PR or to open multiple minor PRs.

Generally, I prefer multiple minor PRs if possible.

File: src/test/java/de/siegmar/logbackgelf/GelfUdpChunkerTest.java [hunk 2]
```diff
@@ -63,4 +64,25 @@ private void expectedChunk(final byte[] data, final int chunkNo, final int chunk
...
     }
 
+    @Test
+    void removeNotPermitted() {
+        final GelfUdpChunker chunker = new GelfUdpChunker(new MessageIdSupplier(), 13);
+        final Iterator<? extends ByteBuffer> chunks =
+                chunker.chunks("hello".getBytes(StandardCharsets.UTF_8)).iterator();
+
+        assertThrows(UnsupportedOperationException.class, chunks::remove);
+    }
```
[the remark this answers is not in the document]
Comment (fabiojb) [reply src/test/java/de/siegmar/logbackgelf/GelfUdpChunkerTest.java:74]: So, can the `ChunkIterator.remove` override be removed, or it needs to be there for compatibility with Java versions bellow 8?

File: src/test/java/de/siegmar/logbackgelf/pool/SimpleObjectPoolTest.java [hunk 3]
```diff
@@ -93,6 +97,79 @@ public void maxLastBorrowed() throws InterruptedException {
...
+
+    @Test
+    void infiniteLifeTime() throws InterruptedException {
+        final SimpleObjectPool<MyPooledObject> pool =
+                new SimpleObjectPool<>(factory, 1, 100, -1, -1);
+
+        final MyPooledObject o1 = pool.borrowObject();
+        pool.returnObject(o1);
+
+        Thread.sleep(200);
```
[the remark this answers is not in the document]
Comment (fabiojb) [reply src/test/java/de/siegmar/logbackgelf/pool/SimpleObjectPoolTest.java:151]: I don't think is really required. I added it as a precaution, to avoid a false positive in case of the implementation changes. I can remove the sleep, or just reduce it to 20 milliseconds. What do you think?

File: src/test/java/de/siegmar/logbackgelf/pool/SimpleObjectPoolTest.java [hunk 4]
```diff
@@ -93,6 +97,79 @@ public void maxLastBorrowed() throws InterruptedException {
...
+
+    @Test
+    void recycleByLifeTime() throws InterruptedException {
+        final SimpleObjectPool<MyPooledObject> pool =
+                new SimpleObjectPool<>(factory, 1, 100, 0, 100);
+
+        final MyPooledObject o1 = pool.borrowObject();
+        pool.returnObject(o1);
+
+        Thread.sleep(2);
```
[the remark this answers is not in the document]
Comment (fabiojb) [reply src/test/java/de/siegmar/logbackgelf/pool/SimpleObjectPoolTest.java:166]: In this case the sleep is required to force the object to be recycled from the pool.

File: src/test/java/de/siegmar/logbackgelf/GelfEncoderTest.java [hunk 5]
```diff
@@ -423,4 +425,80 @@ private String encodeToStr(final LoggingEvent event) {
         return new String(encoder.encode(event), StandardCharsets.UTF_8);
     }
 
+    @Test
+    void staticFieldInvalidFormat() {
```
[the remark this answers is not in the document]
Comment (fabiojb) [reply src/test/java/de/siegmar/logbackgelf/GelfEncoderTest.java]: Done

File: src/test/java/de/siegmar/logbackgelf/GelfEncoderTest.java [hunk 6]
```diff
@@ -423,4 +425,80 @@ private String encodeToStr(final LoggingEvent event) {
...
+    void rewriteStaticField() {
+        encoder.addStaticField("test_id:value");
+        encoder.addStaticField("test_id:new value");
+        assertEquals(1, encoder.getStaticFields().size());
+        assertEquals("value", encoder.getStaticFields().get("test_id"));
+    }
+
+    @ParameterizedTest
+    @NullAndEmptySource
+    void originHostDefaultToLocalHostNameIfEmpty(final String configuredHostname) throws IOException {
```
[the remark this answers is not in the document]
Comment (fabiojb) [reply src/test/java/de/siegmar/logbackgelf/GelfEncoderTest.java:446]: Yes it's not necessary, I didn't realize that. Removed the unnecessary code.

File: src/test/java/de/siegmar/logbackgelf/CustomX509TrustManagerTest.java [hunk 7]
```diff
@@ -104,6 +105,21 @@ public void caSigned() throws Exception {
         validate(cert, caBuilder.getCaCertificate());
     }
 
+    @Test
+    void clientValidation() throws Exception {
+        final X509Util.CABuilder caBuilder = new X509Util.CABuilder();
+        final X509Certificate cert = prepareCaSigned(caBuilder)
+                .build(HOSTNAME);
```
[the remark this answers is not in the document]
Comment (fabiojb) [reply src/test/java/de/siegmar/logbackgelf/CustomX509TrustManagerTest.java]: Fixed

File: src/test/java/de/siegmar/logbackgelf/GelfUdpChunkerTest.java [hunk 8]
```diff
@@ -63,4 +64,25 @@ private void expectedChunk(final byte[] data, final int chunkNo, final int chunk
...
+    void minimumChuckSize() {
+        assertThrows(IllegalArgumentException.class,
+            () -> new GelfUdpChunker(new MessageIdSupplier(), 12));
+    }
+
+    @Test
+    void maximumChunkSize() {
+        assertThrows(IllegalArgumentException.class,
+            () -> new GelfUdpChunker(new MessageIdSupplier(), 65468));
+    }
```
[the remark this answers is not in the document]
Comment (fabiojb) [reply src/test/java/de/siegmar/logbackgelf/GelfUdpChunkerTest.java]: Yep, I changed it to a parameterized test

File: src/test/java/de/siegmar/logbackgelf/pool/SimpleObjectPoolTest.java [hunk 9]
```diff
@@ -93,6 +97,79 @@ public void maxLastBorrowed() throws InterruptedException {
...
+            try {
+                Thread.sleep(200);
+            } catch (InterruptedException e) {
+                throw new RuntimeException(e);
+            }
+            pool.returnObject(o1);
+        }).start();
+
+        final MyPooledObject o2 = pool.borrowObject();
+        final long after = System.currentTimeMillis();
```
[the remark this answers is not in the document]
Comment (fabiojb) [reply src/test/java/de/siegmar/logbackgelf/pool/SimpleObjectPoolTest.java]: Changed to nanoTime

File: src/test/java/de/siegmar/logbackgelf/pool/SimpleObjectPoolTest.java (hunk 1 above)
Comment (fabiojb) [reply src/test/java/de/siegmar/logbackgelf/pool/SimpleObjectPoolTest.java]: Replaced, the code looks way better!

File: src/main/java/de/siegmar/logbackgelf/MessageIdSupplier.java [hunk 10]
```diff
@@ -30,13 +30,14 @@
...
 
-    public int getMachinePart() {
-        return machinePart;
+    private final int machinePart;
+
+    public MessageIdSupplier() {
+        this(RANDOM.nextInt());
     }
 
-    public void setMachinePart(final int machinePart) {
```
[the remark this answers is not in the document]
Comment (fabiojb) [reply src/main/java/de/siegmar/logbackgelf/MessageIdSupplier.java:39]: I'm going to check that.

File: src/test/java/de/siegmar/logbackgelf/GelfUdpChunkerTest.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (osiegmar) [reply src/test/java/de/siegmar/logbackgelf/GelfUdpChunkerTest.java:76]: As the new major version 4 requires Java 8 as minimum, it can safely be removed.

File: src/test/java/de/siegmar/logbackgelf/pool/SimpleObjectPoolTest.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (osiegmar) [reply src/test/java/de/siegmar/logbackgelf/pool/SimpleObjectPoolTest.java:148]: Let's remove it and only add it with a comment on why it is needed (which I doubt).

File: src/test/java/de/siegmar/logbackgelf/pool/SimpleObjectPoolTest.java [hunk 11]
```diff
@@ -122,19 +125,13 @@ void infinitePollWaitTime() throws InterruptedException {
...
-        }).start();
+        Executors.newScheduledThreadPool(1)
+                .schedule(() -> pool.returnObject(o1), 200, TimeUnit.MILLISECONDS);
 
         final MyPooledObject o2 = pool.borrowObject();
-        final long after = System.currentTimeMillis();
+        final long after = System.nanoTime();
 
         assertNotNull(o2);
         assertTrue(after - before > 190);
```
Comment (osiegmar) [inline src/test/java/de/siegmar/logbackgelf/pool/SimpleObjectPoolTest.java:137]: This also needs to be adjusted for nanoTime

**Review by osiegmar — CHANGES_REQUESTED** [review]

File: src/main/java/de/siegmar/logbackgelf/MessageIdSupplier.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (fabiojb) [reply src/main/java/de/siegmar/logbackgelf/MessageIdSupplier.java:39]: I couldn't find any reference showing that is possible set configuration values on java files without a setter. I added the setter again.

File: src/test/java/de/siegmar/logbackgelf/pool/SimpleObjectPoolTest.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (fabiojb) [reply src/test/java/de/siegmar/logbackgelf/pool/SimpleObjectPoolTest.java]: I just removed it

File: src/test/java/de/siegmar/logbackgelf/GelfUdpChunkerTest.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (fabiojb) [reply src/test/java/de/siegmar/logbackgelf/GelfUdpChunkerTest.java:76]: Done

File: src/test/java/de/siegmar/logbackgelf/pool/SimpleObjectPoolTest.java (hunk 11 above)
Comment (fabiojb) [reply src/test/java/de/siegmar/logbackgelf/pool/SimpleObjectPoolTest.java]: Fixed

File: src/main/java/de/siegmar/logbackgelf/MessageIdSupplier.java [hunk 12]
```diff
@@ -30,10 +30,11 @@
...
     // static random to keep Spotbugs happy
     private static final Random RANDOM = new Random();
-    private int machinePart = RANDOM.nextInt();
 
-    public int getMachinePart() {
-        return machinePart;
+    private int machinePart;
+
+    public MessageIdSupplier() {
+        machinePart = RANDOM.nextInt();
```
Comment (osiegmar) [inline src/main/java/de/siegmar/logbackgelf/MessageIdSupplier.java:37]: Are these changes now valuable?

Comment (fabiojb) [reply src/main/java/de/siegmar/logbackgelf/MessageIdSupplier.java]: They're not. Sorry for submitting them, I first reverted manually changing the code. Now I reverted this file to the upstream version using git.

**Comment by osiegmar** [comment]
Thanks again!

**Comment by fabiojb** [comment]
@osiegmar you're welcome! 

Previously you commented that improving the coverage would be a requisite for a new release. Do you have a specific target coverage in mind?

**Comment by osiegmar** [comment]
Very good question. From the pure numbers it's already looking not bad I'd say (not hunting numbers). My current concerns are about the networking stuff (like #62). The corner cases like timeouts or TLS handshakes needs some more testing. And those tests could identify that a different implementation (NIO) is needed.

**Comment by fabiojb** [comment]
Ok, I'll see what I can contribute/test about it. 

About this current implementation, you think that NIO could perform better? Or be more reliable?

**Comment by osiegmar** [comment]
NIO usually comes with higher complexity, hence I'd only use it with a good reason. This reason might be performance – a test has to proof it. The reason might be this socket timeout issue – a test would be needed as well.
